File: code/python/algorithms.py
"""
Shortest-path algorithms: Dijkstra, Bellman-Ford, A*
Pure implementations — no GUI dependencies.
"""

import heapq
import math
from dataclasses import dataclass, field
from typing import Optional

INF = float("inf")

# ─── Graph ───────────────────────────────────────────────────────────────

NODES: dict[str, tuple[int, int]] = {
    "S": (80, 350),  "A": (270, 130), "B": (270, 570),
    "C": (500, 80),  "D": (500, 350), "E": (500, 620),
    "F": (730, 180),  "G": (730, 520), "T": (950, 350),
}

EDGES: list[tuple[str, str, int]] = [
    ("S", "A", 4),  ("S", "B", 2),  ("A", "C", 5),  ("A", "D", 10),
    ("B", "D", 3),  ("B", "E", 8),  ("C", "F", 3),  ("D", "F", 7),
    ("D", "G", 4),  ("E", "G", 2),  ("F", "T", 4),  ("G", "T", 6),
    ("C", "D", 2),
]

ADJ: dict[str, list[tuple[str, int]]] = {n: [] for n in NODES}
for _u, _v, _w in EDGES:
    ADJ[_u].append((_v, _w))
    ADJ[_v].append((_u, _w))

SOURCE = "S"
GOAL = "T"


def heuristic(node: str) -> float:
    x1, y1 = NODES[node]
    x2, y2 = NODES[GOAL]
    return math.hypot(x2 - x1, y2 - y1) / 100


# ─── State ───────────────────────────────────────────────────────────────

@dataclass
class AlgoState:
    dist: dict[str, float] = field(default_factory=dict)
    prev: dict[str, Optional[str]] = field(default_factory=dict)
    visited: set[str] = field(default_factory=set)
    finished: bool = False
    step: int = 0
    message: str = ""
    path: list[str] = field(default_factory=list)
    relaxed_edges: set[tuple[str, str]] = field(default_factory=set)
    current_node: Optional[str] = None
    f_score: dict[str, float] = field(default_factory=dict)
    frontier: list = field(default_factory=list)  # heap for Dijkstra/A*
    bf_iteration: int = 0

    def _reconstruct_path(self) -> None:
        path: list[str] = []
        node: Optional[str] = GOAL
        while node is not None:
            path.append(node)
            node = self.prev.get(node)
        path.reverse()
        self.path = path if path and path[0] == SOURCE else []


# ─── Dijkstra ────────────────────────────────────────────────────────────

def dijkstra_init() -> AlgoState:
    st = AlgoState()
    st.dist = {n: INF for n in NODES}
    st.prev = {n: None for n in NODES}
    st.dist[SOURCE] = 0
    st.frontier = [(0, SOURCE)]
    st.message = f"Dijkstra: start from {SOURCE}, dist[{SOURCE}]=0"
    return st


def dijkstra_step(st: AlgoState) -> None:
    if st.finished or not st.frontier:
        st.finished = True
        st._reconstruct_path()
        c = st.dist.get(GOAL, INF)
        st.message = (f"Dijkstra done! Path: {'→'.join(st.path)}, cost={c:.0f}"
                      if st.path else "Dijkstra done! No path.")
        return

    st.current_node = None
    while st.frontier:
        d, u = heapq.heappop(st.frontier)
        if u in st.visited:
            continue
        st.visited.add(u)
        st.step += 1
        st.current_node = u
        st.message = f"Step {st.step}: visit {u} (dist={st.dist[u]:.0f})"

        for v, w in ADJ[u]:
            if v not in st.visited:
                nd = st.dist[u] + w
                if nd < st.dist[v]:
                    st.dist[v] = nd
                    st.prev[v] = u
                    heapq.heappush(st.frontier, (nd, v))
                    st.relaxed_edges.add((u, v))
                    st.message += f" | relax {u}->{v}:{nd:.0f}"
        return

    st.finished = True
    st._reconstruct_path()
    c = st.dist.get(GOAL, INF)
    st.message = (f"Dijkstra done! Path: {'→'.join(st.path)}, cost={c:.0f}"
                  if st.path else "Dijkstra done! No path.")


# ─── Bellman-Ford ────────────────────────────────────────────────────────

def bellman_ford_init() -> AlgoState:
    st = AlgoState()
    st.dist = {n: INF for n in NODES}
    st.prev = {n: None for n in NODES}
    st.dist[SOURCE] = 0
    st.message = (f"Bellman-Ford: {len(NODES)-1} iterations "
                  f"over {len(EDGES)*2} directed edges")
    return st


def bellman_ford_step(st: AlgoState) -> None:
    if st.finished:
        return

    V = len(NODES)
    all_edges = ([(u, v, w) for u, v, w in EDGES]
                 + [(v, u, w) for u, v, w in EDGES])

    if st.bf_iteration < V - 1:
        changed = False
        for src, dst, w in all_edges:
            if st.dist[src] + w < st.dist[dst]:
                st.dist[dst] = st.dist[src] + w
                st.prev[dst] = src
                st.relaxed_edges.add((src, dst))
                st.visited.add(src)
                st.visited.add(dst)
                changed = True
        st.bf_iteration += 1
        st.step += 1
        st.current_node = None
        label = "changes made" if changed else "no changes (stable)"
        st.message = f"Iteration {st.bf_iteration}/{V-1}: {label}"

        if not changed:
            st.finished = True
            st._reconstruct_path()
            c = st.dist.get(GOAL, INF)
            st.message += (f" -> Early stop! Path: {'→'.join(st.path)}, cost={c:.0f}"
                           if st.path else " -> Early stop!")
    else:
        neg = any(st.dist[s] + w < st.dist[d] for s, d, w in all_edges)
        st.finished = True
        if neg:
            st.message = "Negative cycle detected!"
        else:
            st._reconstruct_path()
            c = st.dist.get(GOAL, INF)
            st.message = (f"Bellman-Ford done! Path: {'→'.join(st.path)}, cost={c:.0f}"
                          if st.path else "Bellman-Ford done! No path.")


# ─── A* ──────────────────────────────────────────────────────────────────

def astar_init() -> AlgoState:
    st = AlgoState()
    st.dist = {n: INF for n in NODES}
    st.prev = {n: None for n in NODES}
    st.f_score = {n: INF for n in NODES}
    st.dist[SOURCE] = 0
    h = heuristic(SOURCE)
    st.f_score[SOURCE] = h
    st.frontier = [(h, SOURCE)]
    st.message = f"A*: start={SOURCE}, goal={GOAL}, h({SOURCE})={h:.1f}"
    return st


def astar_step(st: AlgoState) -> None:
    if st.finished or not st.frontier:
        st.finished = True
        st._reconstruct_path()
        c = st.dist.get(GOAL, INF)
        st.message = (f"A* done! Path: {'→'.join(st.path)}, cost={c:.0f}"
                      if st.path else "A* done! No path.")
        return

    st.current_node = None
    while st.frontier:
        f, u = heapq.heappop(st.frontier)
        if u in st.visited:
            continue
        st.visited.add(u)
        st.step += 1
        st.current_node = u

        if u == GOAL:
            st.finished = True
            st._reconstruct_path()
            st.message = (f"Step {st.step}: GOAL reached! "
                          f"Path: {'→'.join(st.path)}, cost={st.dist[GOAL]:.0f}")
            return

        h = heuristic(u)
        st.message = (f"Step {st.step}: visit {u} "
                      f"(g={st.dist[u]:.0f}, h={h:.1f}, f={f:.1f})")

        for v, w in ADJ[u]:
            if v not in st.visited:
                nd = st.dist[u] + w
                if nd < st.dist[v]:
                    st.dist[v] = nd
                    st.prev[v] = u
                    fv = nd + heuristic(v)
                    st.f_score[v] = fv
                    heapq.heappush(st.frontier, (fv, v))
                    st.relaxed_edges.add((u, v))
                    st.message += f" | relax {u}->{v}: g={nd:.0f}, f={fv:.1f}"
        return

    st.finished = True
    st._reconstruct_path()
    st.message = "A* done! No path."

File: code/csharp/Algorithms.cs
/*
 * Shortest-path algorithms: Dijkstra, Bellman-Ford, A*
 * Pure C# implementations — no GUI dependencies.
 */

using System;
using System.Collections.Generic;
using System.Linq;

namespace ShortestPathViz;

// ─── Graph ──────────────────────────────────────────────────────────────

public record struct Vec2(int X, int Y);
public record struct EdgeDef(string Src, string Dst, int Weight);

public static class Graph
{
    public static readonly Dictionary<string, Vec2> Nodes = new()
    {
        ["S"] = new(80, 350),  ["A"] = new(270, 130), ["B"] = new(270, 570),
        ["C"] = new(500, 80),  ["D"] = new(500, 350), ["E"] = new(500, 620),
        ["F"] = new(730, 180), ["G"] = new(730, 520), ["T"] = new(950, 350),
    };

    public static readonly string[] NodeOrder =
        ["S", "A", "B", "C", "D", "E", "F", "G", "T"];

    public static readonly EdgeDef[] Edges =
    [
        new("S","A",4), new("S","B",2), new("A","C",5), new("A","D",10),
        new("B","D",3), new("B","E",8), new("C","F",3), new("D","F",7),
        new("D","G",4), new("E","G",2), new("F","T",4), new("G","T",6),
        new("C","D",2),
    ];

    public static readonly string Source = "S";
    public static readonly string Goal = "T";

    public static readonly Dictionary<string, List<(string N, int W)>> Adj = BuildAdj();

    private static Dictionary<string, List<(string, int)>> BuildAdj()
    {
        var adj = NodeOrder.ToDictionary(n => n, _ => new List<(string, int)>());
        foreach (var e in Edges)
        {
            adj[e.Src].Add((e.Dst, e.Weight));
            adj[e.Dst].Add((e.Src, e.Weight));
        }
        return adj;
    }

    public static double Heuristic(string node)
    {
        var p = Nodes[node];
        var g = Nodes[Goal];
        return Math.Sqrt(Math.Pow(g.X - p.X, 2) + Math.Pow(g.Y - p.Y, 2)) / 100.0;
    }
}

// ─── State ──────────────────────────────────────────────────────────────

public class AlgoState
{
    public Dictionary<string, double> Dist = new();
    public Dictionary<string, string?> Prev = new();
    public HashSet<string> Visited = new();
    public bool Finished;
    public int Step;
    public string Message = "";
    public List<string> Path = new();
    public HashSet<(string, string)> RelaxedEdges = new();
    public string? CurrentNode;
    public Dictionary<string, double> FScore = new();
    public PriorityQueue<string, double> Frontier = new();
    public int BfIteration;

    public void InitDists()
    {
        foreach (var n in Graph.NodeOrder)
        {
            Dist[n] = double.PositiveInfinity;
            Prev[n] = null;
        }
    }

    public void ReconstructPath()
    {
        Path.Clear();
        string? node = Graph.Goal;
        while (node != null)
        {
            Path.Insert(0, node);
            node = Prev.GetValueOrDefault(node);
        }
        if (Path.Count == 0 || Path[0] != Graph.Source)
            Path.Clear();
    }

    private static string JoinPath(List<string> p) => string.Join("->", p);

    public string PathString() => JoinPath(Path);
}

// ─── Algorithms ─────────────────────────────────────────────────────────

public static class Algorithms
{
    // ── Dijkstra ──

    public static AlgoState DijkstraInit()
    {
        var st = new AlgoState();
        st.InitDists();
        st.Dist[Graph.Source] = 0;
        st.Frontier.Enqueue(Graph.Source, 0);
        st.Message = $"Dijkstra: start from {Graph.Source}";
        return st;
    }

    public static void DijkstraStep(AlgoState st)
    {
        void Finish()
        {
            st.Finished = true;
            st.ReconstructPath();
            st.Message = st.Path.Count > 0
                ? $"Dijkstra done! Path: {st.PathString()}, cost={st.Dist[Graph.Goal]:F0}"
                : "Dijkstra done! No path.";
        }

        if (st.Finished || st.Frontier.Count == 0) { Finish(); return; }

        st.CurrentNode = null;
        while (st.Frontier.Count > 0)
        {
            var u = st.Frontier.Dequeue();
            if (st.Visited.Contains(u)) continue;
            st.Visited.Add(u);
            st.Step++;
            st.CurrentNode = u;
            st.Message = $"Step {st.Step}: visit {u} (dist={st.Dist[u]:F0})";

            foreach (var (v, w) in Graph.Adj[u])
            {
                if (!st.Visited.Contains(v))
                {
                    double nd = st.Dist[u] + w;
                    if (nd < st.Dist[v])
                    {
                        st.Dist[v] = nd;
                        st.Prev[v] = u;
                        st.Frontier.Enqueue(v, nd);
                        st.RelaxedEdges.Add((u, v));
                        st.Message += $" | relax {u}->{v}:{nd:F0}";
                    }
                }
            }
            return;
        }
        Finish();
    }

    // ── Bellman-Ford ──

    public static AlgoState BellmanFordInit()
    {
        var st = new AlgoState();
        st.InitDists();
        st.Dist[Graph.Source] = 0;
        st.Message = $"Bellman-Ford: {Graph.Nodes.Count - 1} iterations over all edges";
        return st;
    }

    public static void BellmanFordStep(AlgoState st)
    {
        if (st.Finished) return;

        int V = Graph.Nodes.Count;
        var allEdges = Graph.Edges
            .SelectMany(e => new[] { (e.Src, e.Dst, e.Weight), (e.Dst, e.Src, e.Weight) })
            .ToList();

        if (st.BfIteration < V - 1)
        {
            bool changed = false;
            foreach (var (s, d, w) in allEdges)
            {
                if (st.Dist[s] + w < st.Dist[d])
                {
                    st.Dist[d] = st.Dist[s] + w;
                    st.Prev[d] = s;
                    st.RelaxedEdges.Add((s, d));
                    st.Visited.Add(s);
                    st.Visited.Add(d);
                    changed = true;
                }
            }
            st.BfIteration++;
            st.Step++;
            st.CurrentNode = null;
            st.Message = $"Iteration {st.BfIteration}/{V - 1}: " +
                         (changed ? "changes made" : "no changes (stable)");

            if (!changed)
            {
                st.Finished = true;
                st.ReconstructPath();
                st.Message += st.Path.Count > 0
                    ? $" -> Early stop! Path: {st.PathString()}, cost={st.Dist[Graph.Goal]:F0}"
                    : " -> Early stop!";
            }
        }
        else
        {
            bool neg = allEdges.Any(e => st.Dist[e.Item1] + e.Item3 < st.Dist[e.Item2]);
            st.Finished = true;
            if (neg)
            {
                st.Message = "Negative cycle detected!";
            }
            else
            {
                st.ReconstructPath();
                st.Message = st.Path.Count > 0
                    ? $"Bellman-Ford done! Path: {st.PathString()}, cost={st.Dist[Graph.Goal]:F0}"
                    : "Bellman-Ford done! No path.";
            }
        }
    }

    // ── A* ──

    public static AlgoState AStarInit()
    {
        var st = new AlgoState();
        st.InitDists();
        foreach (var n in Graph.NodeOrder)
            st.FScore[n] = double.PositiveInfinity;
        st.Dist[Graph.Source] = 0;
        double h = Graph.Heuristic(Graph.Source);
        st.FScore[Graph.Source] = h;
        st.Frontier.Enqueue(Graph.Source, h);
        st.Message = $"A*: start={Graph.Source}, goal={Graph.Goal}, h(S)={h:F1}";
        return st;
    }

    public static void AStarStep(AlgoState st)
    {
        void Finish()
        {
            st.Finished = true;
            st.ReconstructPath();
            st.Message = st.Path.Count > 0
                ? $"A* done! Path: {st.PathString()}, cost={st.Dist[Graph.Goal]:F0}"
                : "A* done! No path.";
        }

        if (st.Finished || st.Frontier.Count == 0) { Finish(); return; }

        st.CurrentNode = null;
        while (st.Frontier.Count > 0)
        {
            var u = st.Frontier.Dequeue();
            if (st.Visited.Contains(u)) continue;
            st.Visited.Add(u);
            st.Step++;
            st.CurrentNode = u;

            if (u == Graph.Goal)
            {
                st.Finished = true;
                st.ReconstructPath();
                st.Message = $"Step {st.Step}: GOAL reached! " +
                             $"Path: {st.PathString()}, cost={st.Dist[Graph.Goal]:F0}";
                return;
            }

            double hu = Graph.Heuristic(u);
            double fu = st.Dist[u] + hu;
            st.Message = $"Step {st.Step}: visit {u} (g={st.Dist[u]:F0}, h={hu:F1}, f={fu:F1})";

            foreach (var (v, w) in Graph.Adj[u])
            {
                if (!st.Visited.Contains(v))
                {
                    double nd = st.Dist[u] + w;
                    if (nd < st.Dist[v])
                    {
                        st.Dist[v] = nd;
                        st.Prev[v] = u;
                        double fv = nd + Graph.Heuristic(v);
                        st.FScore[v] = fv;
                        st.Frontier.Enqueue(v, fv);
                        st.RelaxedEdges.Add((u, v));
                        st.Message += $" | relax {u}->{v}: g={nd:F0}, f={fv:F1}";
                    }
                }
            }
            return;
        }
        Finish();
    }
}
